sdk/src/main/java/com/mapbox/maps/renderer/MapboxWidgetRenderer.kt

```diff
@@ -0,0 +1,171 @@
+    if (textures[0] != 0) {
+      GLES20.glDeleteTextures(textures.size, textures, 0)
+    }
+    GLES20.glGenTextures(1, textures, 0)
+
+    GLES20.glBindTexture(GLES20.GL_TEXTURE_2D, textures[0])
+    GLES20.glTexParameterf(
+      GLES20.GL_TEXTURE_2D,
+      GLES20.GL_TEXTURE_MAG_FILTER,
+      GLES20.GL_LINEAR.toFloat()
+    )
+    GLES20.glTexParameterf(
+      GLES20.GL_TEXTURE_2D,
+      GLES20.GL_TEXTURE_MIN_FILTER,
+      GLES20.GL_LINEAR.toFloat()
+    )
+    GLES20.glTexParameterf(
+      GLES20.GL_TEXTURE_2D,
+      GLES20.GL_TEXTURE_WRAP_S,
+      GLES20.GL_CLAMP_TO_EDGE.toFloat()
+    )
+    GLES20.glTexParameterf(
+      GLES20.GL_TEXTURE_2D,
+      GLES20.GL_TEXTURE_WRAP_T,
+      GLES20.GL_CLAMP_TO_EDGE.toFloat()
+    )
+    GLES20.glTexImage2D(
+      GLES20.GL_TEXTURE_2D,
+      0,
+      GLES20.GL_RGBA,
+      width,
+      height,
+      0,
+      GLES20.GL_RGBA,
+      GLES20.GL_UNSIGNED_BYTE,
+      null
+    )
+    GLES20.glFramebufferTexture2D(
+      GLES20.GL_FRAMEBUFFER, GLES20.GL_COLOR_ATTACHMENT0,
+      GLES20.GL_TEXTURE_2D, textures[0], 0
+    )
+    GLES20.glBindTexture(GLES20.GL_TEXTURE_2D, 0)
+  }
+
+  fun release() {
+    if (widgetEglPrepared) {
+      widgetEglPrepared = false
+
+      if (widgetEglSurface != widgetEglCore.EGL_NO_SURFACE) {
+        widgetEglCore.makeCurrent(widgetEglSurface)
+
+        GLES20.glDeleteFramebuffers(framebuffers.size, framebuffers, 0)
+        GLES20.glDeleteTextures(textures.size, textures, 0)
+        widgets.forEach {
+          it.renderer.deinitialize()
+        }
+        widgets.clear()
+
+        widgetEglCore.releaseSurface(widgetEglSurface)
+
+        widgetEglSurface = widgetEglCore.EGL_NO_SURFACE
+      }
+
+      widgetEglCore.release()
+    }
+  }
+
+  fun updateTexture() {
+    if (needRender) {
+      initEgl()
+      widgetEglCore.makeCurrent(widgetEglSurface)
+      GLES20.glBindFramebuffer(GLES20.GL_FRAMEBUFFER, framebuffers[0])
+      textureToFramebuffer()
+      widgets.forEach {
+        it.renderer.render()
+      }
+      GLES20.glBindFramebuffer(GLES20.GL_FRAMEBUFFER, 0)
+      widgetEglCore.makeNothingCurrent()
+    }
+  }
+
+  private fun initEgl() {
+    if (widgetEglPrepared && widgetEglSurface != widgetEglCore.EGL_NO_SURFACE) {
+      return
+    }
+
+    if (!widgetEglPrepared) {
+      widgetEglCore.prepareEgl()
+      if (!widgetEglCore.eglStatusSuccess) {
+        Logger.e(TAG, "Widget EGL was not configured, please check logs above.")
+        return
+      }
+      widgetEglPrepared = true
+    }
+
+    if (widgetEglSurface == widgetEglCore.EGL_NO_SURFACE) {
+      widgetEglSurface =
+        widgetEglCore.createOffscreenSurface(width = width, height = height)
+      if (widgetEglSurface == widgetEglCore.EGL_NO_SURFACE) {
+        return
+      }
+    }
+
+    widgetEglCore.makeCurrent(widgetEglSurface)
+
+    GLES20.glGenFramebuffers(1, framebuffers, 0)
+    GLES20.glBindFramebuffer(GLES20.GL_FRAMEBUFFER, framebuffers[0])
```

Comment:
Looks pretty strange - exiting function and not clearing framebuffer.
I guess usual code should look like:
```
        GLES20.glGenFramebuffers(1, values, 0)
        val frameBufferId = values[0]
        GLES20.glBindFramebuffer(GLES20.GL_FRAMEBUFFER, frameBufferId)
        GLES20.glFramebufferTexture2D(
            GLES20.GL_FRAMEBUFFER, GLES20.GL_COLOR_ATTACHMENT0,
            GLES20.GL_TEXTURE_2D, textureId, 0
        )
        val status = GLES20.glCheckFramebufferStatus(GLES20.GL_FRAMEBUFFER)
        if (status != GLES20.GL_FRAMEBUFFER_COMPLETE) {
            throw RuntimeException("Framebuffer not complete, status=$status")
        }
        GLES20.glBindFramebuffer(GLES20.GL_FRAMEBUFFER, 0)
```

Reply:
not precisely this code but guess you got my idea

Reply:
Reworked